cmd/snap-bootstrap/bootstrap/tpm.go

```diff
@@ -0,0 +1,205 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package bootstrap
+
+import (
+	"crypto/rand"
+	"fmt"
+	"io/ioutil"
+
+	"github.com/chrisccoulson/go-tpm2"
+	"github.com/snapcore/secboot"
+
+	"github.com/snapcore/snapd/osutil"
+)
+
+const (
+	// Handles are in the block reserved for owner objects (0x01800000 - 0x01bfffff)
+	pinHandle = 0x01800000
+)
+
+type tpmSupport struct {
+	// Connection to the TPM device
+	tconn *secboot.TPMConnection
+	// Lockout authorization
+	lockoutAuth []byte
+	// List of paths to shim files
+	shimFiles []string
+	// List of paths to bootloader files
+	bootloaderFiles []string
+	// List of paths to kernel files
+	kernelFiles []string
+}
+
+func NewTPMSupport() (*tpmSupport, error) {
+	tconn, err := secboot.ConnectToDefaultTPM()
+	if err != nil {
+		return nil, fmt.Errorf("cannot connect to default TPM: %v", err)
+	}
+
+	lockoutAuth := make([]byte, 16)
+	// crypto rand is protected against short reads
+	_, err = rand.Read(lockoutAuth)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create lockout authorization: %v", err)
+	}
+
+	t := &tpmSupport{tconn: tconn, lockoutAuth: lockoutAuth}
+
+	return t, nil
+}
+
+// StoreLockoutAuth saves the lockout authorization data in a file at the
+// path specified by filename.
+func (t *tpmSupport) StoreLockoutAuth(filename string) error {
+	if err := ioutil.WriteFile(filename, t.lockoutAuth, 0600); err != nil {
+		return err
+	}
+	return nil
+}
+
+// SetShimFiles verifies and sets the list of shim binaries.
+func (t *tpmSupport) SetShimFiles(pathList ...string) error {
```

Comment:
Silly question - right now we only use all of this for a single file not for multiple files, should we add support for multiple files when we need it? Or will it be needed really really soon so there is no win in simplifying this to a single shim/bootloader/kernel?

Reply:
Not accepting multiple files would simplify the nested loops in `tpmSupport.Seal()` but these would be eventually needed to be added back to deal with key rotation. I would keep the code since it's already there, but I can replace it with a simpler version if you feel it's more appropriate at this moment.

Reply:
See comment below for the `TestSeal()`